src/server/db_slice.cc

```diff
@@ -1338,4 +1361,57 @@ void DbSlice::ResetUpdateEvents() {
   events_.update = 0;
 }
 
+// todo: perhaps we need to limit the total number of entry in the tracking
+// table so that we have a way to limit the amount of memory used for
+// client tracking
+void DbSlice::TrackKeys(ConnectionContext* cntx, int32_t tid,
+                        const std::vector<std::string_view>& keys) {
+  DVLOG(2) << "Start tracking the following keys for thread ID: " << tid
+           << ", client ID: " << cntx->conn()->GetClientId();
+  for (auto key : keys) {
+    // std::string_view k = key;
+    std::string k{key.begin(), key.end()};
+    DVLOG(2) << "  " << k;
+    if (client_tracking_map_.find(k) == client_tracking_map_.end()) {
+      std::pair<ConnectionContext*, int32_t> p{cntx, tid};
+      absl::flat_hash_set<std::pair<ConnectionContext*, int32_t>, Hash> tracker_set{p};
+      client_tracking_map_.insert({k, tracker_set});
+    } else {
+      std::pair<ConnectionContext*, int32_t> p{cntx, tid};
+      client_tracking_map_[k].insert(p);
+    }
+  }
+}
+
+void DbSlice::SendInvalidationTrackingMessage(std::string_view key) {
+  std::string k{key.begin(), key.end()};
+  if (client_tracking_map_.find(k) != client_tracking_map_.end()) {
+    // notify all the clients.
+    auto& client_set = client_tracking_map_[k];
+    DVLOG(2) << "Garbage collect clients that are no longer tracking... ";
+    auto is_not_tracking = [](std::pair<ConnectionContext*, int32_t> p) {
+      return (!p.first->conn()->IsTrackingOn());
+    };
+    absl::erase_if(client_set, is_not_tracking);
```

Comment:
The client should unregister itself from `Service::OnClose()`

Reply:
currently I'm relying on a lazy garbage collection of those closed clients/contexts in the table: when a key is updated, its entry in the tracking table will be removed, and therefore, all the clients that are/were tracking this key will be removed from the table. 

It could be very costly to remove the client from tracking table especially when there are many keys are being tracked. (we have to iterate through the table to check if the client is involved there)

Reply:
> Then:
> 
> 1. You have to use Connection::WeakRef if you want to store the pointers indefinitely long

Yes, that's the plan: Use WeakRef to replace the raw pointer to the context. 

> 2. You can only check IsTrackingOn on the destination thread (or you make this field atomic_bool, but then the pointer problem becomes more difficult)

Yes, that's the current garbage collection logic in the implementation. As long as the WeakRef is a valid pointer, it will allow me to access the tracking flag which will be set to false.

> 3. You can still accumulate lot's of outdated entries for rarely used keys

That's true. Note that Redis currently uses a very similar way as well. Certainly one can implement another background garbage collection in heartbeat to mitigate this.